File: 2019/day_01/solution.py
with open('data.txt', 'r') as puzzle:
    modules = [int(e) for e in puzzle.read().split('\n') if e]

def fuel(mass: int) -> int:
    return max(mass // 3 - 2, 0)

def fuel_rec(module: int, acc: int = 0) -> int:
    f = fuel(module)
    if f:
        return fuel_rec(f, acc+f)
    else:
        return acc


print("part 1: ", sum([fuel(e) for e in modules]))
print("part 2: ", sum([fuel_rec(e) for e in modules]))

File: 2019/day_01/solution.exs
defmodule Day1 do
  def fuel(x) do
    max(div(x, 3) - 2, 0)
  end

  def fuel_rec(x, acc) do
    case fuel(x) do
      0 -> acc
      f -> fuel_rec(f, acc+f)  
    end
  end
end

modules = File.read!("data.txt")
  |> String.split
  |> Enum.map(&String.to_integer/1)


IO.puts Day1.fuel_rec(14, 0)
IO.puts "part 1: #{Enum.map(modules, fn x -> Day1.fuel(x) end) |> Enum.sum}"
IO.puts "part 2: #{Enum.map(modules, fn x -> Day1.fuel_rec(x, 0) end) |> Enum.sum}"